"struct defined here" doesn't show where the struct is defined

<!--
Thank you for filing a bug report! 🐛 Please provide a short summary of the bug,
along with any information you feel relevant to replicating the bug.

If you cannot produce a minimal reproduction case (something that would work in
isolation), please provide the steps or even link to a repository that causes
the problematic output to occur.
-->

When a `struct` defined outside of the current crate is used with not enough generics, the note in the error says "struct defined here" but doesn't show where.

Given the following [code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=561e2845762fff0f305fcbdab454916d): <!-- Please provide a link to play.rust-lang.org -->

```rust
struct S<T>(T);

pub fn f(_: Box, _: S)  {}
```

The current output is:

```
error[E0107]: missing generics for struct `Box`
   --> src/lib.rs:3:13
    |
3   | pub fn f(_: Box, _: S)  {}
    |             ^^^ expected at least 1 generic argument
    |
note: struct defined here, with at least 1 generic parameter: `T`
help: add missing generic argument
    |
3   | pub fn f(_: Box<T>, _: S)  {}
    |             ~~~~~~

error[E0107]: missing generics for struct `S`
 --> src/lib.rs:3:21
  |
3 | pub fn f(_: Box, _: S)  {}
  |                     ^ expected 1 generic argument
  |
note: struct defined here, with 1 generic parameter: `T`
 --> src/lib.rs:1:8
  |
1 | struct S<T>(T);
  |        ^ -
help: add missing generic argument
  |
3 | pub fn f(_: Box, _: S<T>)  {}
  |                     ~~~~
```

<!-- The following is not always necessary. -->
Ideally, the output should look like this:

```
error[E0107]: missing generics for struct `Box`
   --> ./test.rs:3:13
    |
3   | pub fn f(_: Box, _: S) {}
    |             ^^^ expected at least 1 generic argument
    |
note: struct defined here, with at least 1 generic parameter: `T`
   --> /.../repos/rust/library/alloc/src/boxed.rs:172:12
    |
172 | pub struct Box<
    |            ^^^
173 |     T: ?Sized,
    |     -
help: add missing generic argument
    |
3   | pub fn f(_: Box<T>, _: S) {}
    |             ~~~~~~

# error for S unchanged
```

<!--
If the problem is not self-explanatory, please provide a rationale for the
change.
-->

<!--
If dramatically different output is caused by small changes, consider also
adding them here.

If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions. The output might also be
different depending on the Edition.
-->

---

I'm willing to work on this, though I don't have a clue why this is happening (yet). An interesting thing to notice is that the current box error is actually indented more, I suppose it's because `Box`  is located on a 3 digit line (172, currently). So it's like that part of the error is somewhere, but it's lost?

--- a/compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs
+++ b/compiler/rustc_typeck/src/structured_errors/wrong_number_of_generic_args.rs
@@ -731,7 +731,11 @@ impl<'a, 'tcx> WrongNumberOfGenericArgs<'a, 'tcx> {
     /// Builds the `type defined here` message.
     fn show_definition(&self, err: &mut DiagnosticBuilder<'_>) {
         let mut spans: MultiSpan = if let Some(def_span) = self.tcx.def_ident_span(self.def_id) {
-            def_span.into()
+            if self.tcx.sess.source_map().span_to_snippet(def_span).is_ok() {
+                def_span.into()
+            } else {
+                return;
+            }
         } else {
             return;
         };

